lib/java/persistance/src/main/java/org/enso/persist/Persistance.java

```diff
@@ -259,4 +263,72 @@ public static <V> Reference<V> of(V obj) {
       return new PerMemoryReference<>(obj);
     }
   }
+
+  /**
+   * Reference to an object.
+   *
+   * <p>This is a sibling to {@link Reference} that has the same lazy behaviour upon read, but it is
+   * written eagerly. This class is not meant to be used when cycles in the structure are expected -
+   * use {@link Reference} instead. It should be used when we want to ensure that a part of a
+   * structure is only read lazily.
+   */
+  public abstract static sealed class InlineReference<T>
+      permits DirectInlineReference, IndirectInlineReference {
```

Comment:
My rationale for making this a distinct type from `Reference`:

The property of being inline or deferred is now a property of the 'slot' (determined by the class when reading/writing the object) and not by the actual 'value'. I have slots in my structure which I want to be inline and ones that need to be ready for circular reference, thus need to be deferred.

I'm not sure if that property could be moved to be property of the 'value'. Maybe. But for me it was conceptually **much** simpler to think of this a as a property of the 'slot' - i.e. the expected type.

It also keeps the `Reference` special cases relatively simple, and the logic for `InlineReference` is nicely encapsulated in `PersistInlineReference`. I think that follows the Single Responsibility Principle.

Reply:
...just the API is ugly...

Reply:
As requested in DM, I split off the Persistance changes into a separate PR: https://github.com/enso-org/enso/pull/10101